### PR TITLE
feat(@astrojs/netlify): Add on-demand builders Netlify functions 

### PR DESCRIPTION
**Pull request by juanmiguelguerrero**
## Changes

- A new configuration option is added to the Netlify adapter to enable On-Demand Builder feature.

## Testing

- I don't know how to add a test with `chai`.
- In any case, by adding this configuration, it is possible to test the functionality:

```javascript
import { defineConfig } from 'astro/config';
import netlify from '@astrojs/netlify/functions';

export default defineConfig({
  output: 'server',
  adapter: netlify({
    builders: true
  }),
});

```

## Docs

- A new heading "On-Demand Builders" must be added to the "Configuration" section, including the above example and a link to the official documentation (https://docs.netlify.com/configure-builds/on-demand-builders/)

/cc @withastro/maintainers-docs for feedback!

### CODE REVIEW

**Comment by matthewp** [comment]
Very nice! Can you add a changeset? `pnpm changeset`, this would be a `minor` change since it's a new feature.

**Comment by juanmiguelguerrero** [comment]
@matthewp There is a CI failure on macOS. It's related to a timeout in a Astro CLI test . I think this should not be impacted by this pull request. Thanks.

**Comment by matthewp** [comment]
@juanmiguelguerrero Yeah thanks, I believe that's a flaky test. I'll rerun.

**Comment by juanmiguelguerrero** [comment]
I found an error in the text example.

This is the correct code:

```javascript
import { defineConfig } from 'astro/config';
import netlify from '@astrojs/netlify/functions';

export default defineConfig({
  output: 'server',
  adapter: netlify({
    builders: true
  }),
});

```

Note: I have edited the previous comment.

**Review by matthewp — APPROVED** [review]

**Comment by sarah11918** [comment]
@juanmiguelguerrero Here is the file that needs to be edited to add something to the docs `configuration-reference` page:  https://github.com/withastro/astro/blob/main/packages/astro/src/%40types/astro.ts

Can you please document the necessary changes there, following the style of the other items at https://docs.astro.build/en/reference/configuration-reference/ ? Let me know if you need/want assistance with this.

Then the docs team will review them!

**Comment by juanmiguelguerrero** [comment]
@sarah11918 I think there is nothing to change in that file.

This code changes do not affect the general Astro configuration. They only affect the Netlify adapter configuration.

The doc to update is this one: https://docs.astro.build/en/guides/integrations-guide/netlify/

https://github.com/withastro/docs/blob/099aba24730af21f81770302c0a6796163b11c4c/src/pages/en/guides/integrations-guide/netlify.mdx

In my opinion, the place to insert this change is inside the "Configuration" section, between the "dist" and "binaryMediaTypes" options.

capture

The text we should include would be something like this:

### builders

On-demand Builders are serverless functions used to generate web content as needed that’s automatically cached on Netlify’s Edge CDN. To activate rendering with this type of functions, use the `builders` option:

```javascript
import { defineConfig } from 'astro/config';
import netlify from '@astrojs/netlify/functions';

export default defineConfig({
  output: 'server',
  adapter: netlify({
    builders: true
  }),
});

```

This functionality is only available whith the `@astrojs/netlify/functions` adapter and is not compatible with Edge Functions.

For more information about Netlify On-demand Builders visit the [Netlify documentation.](https://docs.netlify.com/configure-builds/on-demand-builders/)

**Comment by sarah11918** [comment]
@juanmiguelguerrero Gotcha! In that case, you'll want to edit this page: https://github.com/withastro/astro/blob/main/packages/integrations/netlify/README.md 

This content is pulled into the Docs site from this README. Would you mind opening a PR with proposed changes there? Docs will get notification of that PR automatically, and that's where we can finalize the wording together.  Thanks!

**Comment by juanmiguelguerrero** [comment]
@sarah11918 I have just updated the documentation.

**Comment by juanmiguelguerrero** [comment]
@matthewp Jhon, can you resolve these conflicts? 

@sarah11918 Have you seen the changes I made to the README? Are they correct?

Can I help? Thanks... 🚀

File: packages/integrations/netlify/README.md [hunk 1]
```diff
@@ -117,6 +117,28 @@ And then point to the dist in your `netlify.toml`:
...
+  output: 'server',
+  adapter: netlify({
+    builders: true
+  }),
+});
+```
+
+This functionality is only available whith the `@astrojs/netlify/functions` adapter and is not compatible with Edge Functions.
+
+For more information about Netlify On-demand Builders visit the [Netlify documentation.](https://docs.netlify.com/configure-builds/on-demand-builders/)
```
Comment (sarah11918) [inline packages/integrations/netlify/README.md:139]:
```suggestion
```
I think we can replace this with an inline link above, which has the advantage of using more descriptive link text ("Netlify On-demand Builders" vs "Netlify Documentation").  If you'd rather keep this sentence instead (we don't need both links!), then I would change the linked part to  be more descriptive and include "Netlify On-demand Builders", something like:

> See more in [Netlify's documentation for On-demand Builders](https://docs.netlify.com/configure-builds/on-demand-builders/).

**Review by sarah11918 — COMMENTED** [review]
Thanks for the ping to let me know this was updated! Here are my suggestions for documentation text, if they don't lose your original intention!

File: packages/integrations/netlify/README.md [hunk 2]
```diff
@@ -117,6 +117,28 @@ And then point to the dist in your `netlify.toml`:
 directory = "dist/functions"
 ```
 
+### builders
+
+On-demand Builders are serverless functions used to generate web content as needed that’s automatically cached on Netlify’s Edge CDN. To activate rendering with this type of functions, use the `builders` option:
```
[the remark this answers is not in the document]
Comment (juanmiguelguerrero) [reply packages/integrations/netlify/README.md:122]: Perfect!!

File: packages/integrations/netlify/README.md (hunk 1 above)
Comment (juanmiguelguerrero) [reply packages/integrations/netlify/README.md]: Ok

File: packages/integrations/netlify/README.md [hunk 3]
```diff
@@ -117,6 +117,28 @@ And then point to the dist in your `netlify.toml`:
...
+
+export default defineConfig({
+  output: 'server',
+  adapter: netlify({
+    builders: true
+  }),
+});
+```
+
+This functionality is only available whith the `@astrojs/netlify/functions` adapter and is not compatible with Edge Functions.
```
[the remark this answers is not in the document]
Comment (juanmiguelguerrero) [reply packages/integrations/netlify/README.md:137]: I think this is correct

File: .changeset/twenty-pans-agree.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@astrojs/netlify': minor
+---
+
+Add on-demand builders Netlify functions
```
Comment (sarah11918) [inline .changeset/twenty-pans-agree.md:5]:
```suggestion
Add `builders` config option for Netlify On-demand Builders.
```
Just a suggestion! Netlify does appear to capitalize both "On" and "Builders". And, the change is adding this new config option to the adapter, so this is explicit about that. (Don't think you need to say that it's only for Netlify Functions, not Edge, in this changelog description. That's for the README.)

**Comment by juanmiguelguerrero** [comment]
@matthewp I have already reviewed the pending conflicts.

**Comment by matthewp** [comment]
Thanks @juanmiguelguerrero! This is looking great.

@sarah11918 just need your approval if there's nothing else.

**Review by sarah11918 — APPROVED** [review]
Docs is happy!  Thanks @juanmiguelguerrero  :raised_hands:

**Comment by betabong** [comment]
Netlify on-demand builders allow setting a Time to Live duration – this can be very useful for content like e-commerce or larger CMS setups. Ideally this could be configured per route (`Astro.response`). I was hoping that setting a header could accomplish it, but it seems like Netlify expects to have this set in the response. So may be "builders" could become a more complex option?

a. `builder: true`
b. `builder: 3600`
c. `builder: (event, context) => event.path.startsWith("/archive") ? false : 60`

**Comment by juanmiguelguerrero** [comment]
@betabong It is an interesting functionality and is not difficult to implement.

My only doubt is how would be the best configuration option?

I think the simplest is something like this:

```javascript
import { defineConfig } from 'astro/config';
import netlify from '@astrojs/netlify/functions';

export default defineConfig({
  output: 'server',
  adapter: netlify({
    builders: true,
    ttl: 3600
  }),
});

```

Anyway I would like to know before the opinion of @matthewp

**Comment by betabong** [comment]
Hey @juanmiguelguerrero I have made a pull request here https://github.com/withastro/astro/pull/6219   

Please have a look, thanks 🙏

**Pull request merged**